pkg/logql/log/dedup_filter.go

```diff
@@ -0,0 +1,59 @@
+package log
+
+import (
+	"github.com/prometheus/prometheus/pkg/labels"
+)
+
+type LineDedupFilter struct {
+	labels     map[string]interface{}
+	inverted   bool
+	hashLookup map[uint64]struct{}
+}
+
+func NewLineDedupFilter(labelFilters []string, inverted bool) *LineDedupFilter {
+	// create a map of labelFilters for O(1) lookups instead of O(n)
+	var filterMap = make(map[string]interface{})
+	for _, group := range labelFilters {
+		filterMap[group] = nil
+	}
+
+	return &LineDedupFilter{
+		labels:     filterMap,
+		inverted:   inverted,
+		hashLookup: make(map[uint64]struct{}),
+	}
+}
+
+func (l *LineDedupFilter) Process(line []byte, lbs *LabelsBuilder) ([]byte, bool) {
+	var filterLabels labels.Labels
+
+	for _, label := range append(lbs.base, lbs.add...) {
```

Comment:
I think this should use `lbs.Labels()`as it encapsulates all the logic necessary (including unmentioned deletions). @cyriltovena Am I correct in this assumption? It would be nice to avoid unnecessary allocs, but I'm unsure the correctness guarantees hold.

Reply:
Yes !

Reply:
Ah nice, implemented